src/imp/schannel.rs

```diff
@@ -140,15 +154,126 @@ impl Identity {
         }
         Ok(Identity { cert: context })
     }
+
+    pub fn from_os_provider(_pem: &[u8], provider_name: &OsStr, os_engine_string: &OsStr) -> Result<Identity, Error> {
+        if provider_name != "ncrypt" && provider_name != "e_ncrypt" {
+            return Err(io::Error::new(io::ErrorKind::InvalidInput,"`provider_name` must be either ncrypt or e_ncypt").into())
+        } 
+
+        let os_provider = parse_engine_string(os_engine_string)
+        .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput,"Invalid `os_engine_string`"))?;
+
+        match os_provider {
+            OsProviderParameters::ContextFromStore {store_name, is_machine, decoded_hex} => {
+
+                let store = if is_machine {CertStore::open_local_machine(&store_name)} else {CertStore::open_current_user(&store_name)}?;
+                let mut identity = None;
+            
+                for cert in store.certs() { 
+
+                    let algo: HashAlgorithm;
+                    match decoded_hex.len() {
+                        20 => algo = HashAlgorithm::sha1(), 
+                        32 => algo = HashAlgorithm::sha256(), 
+                        _ => return Err(io::Error::new(io::ErrorKind::InvalidInput,"Invalid hex thumbprint").into())
+                    }
+
+                    let hash = cert.fingerprint(algo)?; 
+                    if hash == decoded_hex { 
+                        if cert 
+                            .private_key() 
+                            .silent(true) 
+                            .compare_key(true) 
+                            .acquire() 
+                            .is_err() 
+                        { 
+                            return Err(io::Error::new(io::ErrorKind::InvalidInput,"Missing or invalid private key property").into())
+                        } 
+                        identity = Some(cert);
+                        break; 
+                    } 
+                } 
+
+                let identity = match identity {
```

Comment:
can use `let identity = identity.ok_or_else(...)?;` or even better replace this and the return expr with `identity.ok_or_else(...).map(...)`